### PR TITLE
Support WASB scheme in ADLSFileIO

### PR DESCRIPTION
**Pull request by mrcnc**
This is a second attempt to resolve https://github.com/apache/iceberg/issues/10127 but avoids using `java.net.URI` for parsing, which was found to be problematic after the first attempt was merged.

Additionally I've refactored to minimize the number of lines changed and clarify the usage of the `storageAccount` variable, which was previously storing the endpoint host and now will only store the [storage account name](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-overview#standard-endpoints) as the subdomain.

### CODE REVIEW

File: azure/src/main/java/org/apache/iceberg/azure/AzureProperties.java [hunk 1]
```diff
@@ -93,7 +93,7 @@ public void applyClientConfiguration(String account, DataLakeFileSystemClientBui
     if (connectionString != null && !connectionString.isEmpty()) {
       builder.endpoint(connectionString);
     } else {
-      builder.endpoint("https://" + account);
+      builder.endpoint("https://" + account + ".dfs.core.windows.net");
```
Comment (jbonofre) [inline azure/src/main/java/org/apache/iceberg/azure/AzureProperties.java:96]: Should we really change add `.def.core.windows.net` by default ?
I mean that use currently include the suffix in `account` so it should be clearly state that we append the suffix by default now.

**Review by jbonofre — APPROVED** [review]
LGTM, I just wonder about the suffix appended by default to `account`.

File: azure/src/main/java/org/apache/iceberg/azure/AzureProperties.java (hunk 1 above)
Comment (RussellSpitzer) [reply azure/src/main/java/org/apache/iceberg/azure/AzureProperties.java:96]: I think this is the right thing to do, I would just throw a javadoc on this method though

**Review by RussellSpitzer — APPROVED** [review]
@bryanck + @danielcweeks - Please check

File: azure/src/main/java/org/apache/iceberg/azure/AzureProperties.java (hunk 1 above)
Comment (bryanck) [reply azure/src/main/java/org/apache/iceberg/azure/AzureProperties.java:107]: This deviates from the `abfs[s]` scheme defined by Hadoop, where the domain is specified in the URI ([docs](https://hadoop.apache.org/docs/stable/hadoop-azure/abfs.html)). I feel we should check if the account already ends with the domain and only append then to support that.

File: azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSLocation.java [hunk 2]
```diff
@@ -30,14 +30,21 @@
  *
  * <p>Locations follow a URI like structure to identify resources
  *
- * <pre>{@code abfs[s]://[<container>@]<storage account host>/<file path>}</pre>
+ * <pre>{@code abfs[s]://[<container>@]<storageAccount>.dfs.core.windows.net/<path>}</pre>
```
Comment (bryanck) [inline azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSLocation.java:33]: I feel we should stick to the URI format consistent w/ Hadoop's definition (without assuming the domain)

**Review by bryanck — APPROVED** [review]

**Review by amogh-jahagirdar — APPROVED** [review]
Thanks @mrcnc , yeah I think supporting WASB despite it being deprecated by Azure is worth it and I think the way this PR handles it is correct.

**Comment by bryanck** [comment]
Thanks for the PR @mrcnc ! And for the reviews @RussellSpitzer @amogh-jahagirdar @jbonofre !

**Comment by ajreid21** [comment]
This unfortunately is/was a breaking change 😞 - related to the concern @jbonofre brought up.

**Comment by ajreid21** [comment]
AzureProperties builds a map of account -> sasToken here when you create ADLSFileIO using `adls.sas-token.` as the credential mechanism. 

Prior to this change, the account passed in [here](https://github.com/apache/iceberg/blob/main/azure/src/main/java/org/apache/iceberg/azure/AzureProperties.java#L91) (which is used to lookup the sasToken from the map mentioned above) was parsed to include `.dfs.core.windows.net` so when generating the `adls.sas-token.` property to pass into ADLSFileIO, you needed to include `.dfs.core.windows.net` as part of the `adls.sas-token.` property name. After this change, we are parsing the ADLSLocation account to NOT include `adls.sas-token.`, so now the sasToken lookup from the map doesn't find the sasToken (unless the user updates their code to not include it in the property name anymore)...

**Comment by bryanck** [comment]
In retrospect this is probably a feature change we shouldn't have added in a patch release. We should probably revert this and have another patch release.

**Pull request merged**